include/toolchain/common.h

```diff
@@ -180,11 +180,22 @@
  * Convenience helper combining __in_section() and Z_DECL_ALIGN().
  * The section name is the struct type prepended with an underscore.
  * The subsection is "static" and the subsubsection is the variable name.
+ *
+ * In the linker script, create output sections for these using
+ * Z_ITERABLE_SECTION_ROM or Z_ITERABLE_SECTION_RAM.
  */
 #define Z_STRUCT_SECTION_ITERABLE(struct_type, name) \
 	Z_DECL_ALIGN(struct struct_type) name \
 	__in_section(_##struct_type, static, name) __used
 
+/* Special variant of Z_STRUCT_SECTION_ITERABLE, for placing alternate
+ * data types within the iterable section of a specific data type. The
+ * data type sizes and semantics must be equivalent!
+ */
+#define Z_STRUCT_SECTION_ITERABLE_ALTERNATE(out_type, struct_type, name) \
```

Comment:
Seems like we should be able to fix the naming and init conventions for sem and queue/fifo/lifo such that this isn't needed, but it's a 2-line internal API so maybe not worth worrying about.